text/0779-first-class-component-templates.md

```diff
@@ -0,0 +1,1613 @@
+[tree-sitter-highlighting]: https://github.com/alexlafroscia/tree-sitter-glimmer
+
+
+#### Blueprints
+
+All blueprints will need to be updated to support generating the new format. During the transition period, we should allow generating both. The rollout will follow the example of the rollout of Glimmer components with Octane:
+
+1. Introduce the ability to author components in the new format with a new `--strict` flag, but leave the default today’s loose mode format. Introduce `--loose` as an explicit flag for using today’s loose mode format.
+
+2. When Ember Polaris[^polaris] is released, change the default to the new format for apps and addons which set `"edition": "polaris"`, while leaving loose mode available via `--loose`, and preserving `--strict` as an explicit flag for the new default.
+
+3. If or when loose mode templates are deprecated, the supporting blueprint infrastructure can be removed, including the `--loose` flag.
+
+The current blueprints support generating a backing class for any existing component template which does not already have a backing class with the `component-class` format. We have two choices about the behavior of that blueprint for strict mode templates:
+
+1. Do not support it. Adding a backing class is simply a matter of adding an import and adding a class.
+
+2. Re-implement the blueprint using an AST transform (which we have prior art for: route generation uses that approach), to add a backing class for an existing default export in the module.
+
+(We should take (1) here as a default starting point, while encouraging the community to implement (2) if interested.)
+
+We should also update the name of the class generated for a component class. The default behavior of today's blueprint when generating a component is to suffix the class name with `Component`. Thus, running `ember generate component greeting --component-class=@glimmer/component` will produce a class named `GreetingComponent`.[^ts-component-name]
+
+There was room for debate about whether this made sense for naming component classes up till now, since the invocation name was based on the file name (using Ember's resolution rules) and *not* the class name. Now, though, it *will* be based on the imported name, and the standard behavior of auto-import tooling is to import classes by their full name—whether the item is a named export or a default export. When a user goes to auto-complete `Greeting` (e.g. in [Glint][glint]), they will end up with `GreetingComponent`, leading to this sort of thing if they don’t rename it:
+
+```js
+import GreetingComponent from './greeting';
+
+<template>
+  <GreetingComponent @name={{@user.name}} />
+</template>
+```
+
+This is obviously undesirable, but avoiding this will mean mean renaming locally after the auto-complete works. That renaming operation is a needless paper cut in the best case of importing a default export. It rises to the level of a significant annoyance when using named imports:
```

Comment:
A counterpoint here is that this new world _easily_ enables multiple exports of various types and that disambiguating which are components vs helpers vs modifiers making the suffix possibly more important.

Thoughts?

Reply:
This is a good question. First, it’s worth noting that this kind of disambiguation *is* conceivably necessary, because having multiple items of the same name in a module is a `SyntaxError`. Assuming there might be cases where you want multiple items of the same *base* name, suffixing to disambiguate could be useful—especially since `{{thisInvocation}}` could be a helper invocation *or* a curly-component invocation of a component.

That all being said, I think that the *default* should be no-suffix, because while the ambiguities you describe exist, they aren’t the *norm*:

1. Needing to disambiguate *by name* between helpers and curly components like this is exceedingly rare in Octane apps, which generally use angle-bracket invocation.

2. In most of the rare cases where the possibility of ambiguity between a helper and a curly component invocation *does* appear, it is more or less irrelevant whether it is a helper or a component: the same invocation rules apply either way. (I am actually trying to come up with a scenario where a user would care *and* wouldn't be using curly invocation, and failing, but I'd be happy to admit such an example!)

3. Of *those* cases, it is even less common that the items of the same base name would be coming from the same module. (I would personally take it as a pretty significant failure of naming design, and would absolutely reject it in a code review!)

4. For those remaining cases, one of the great advantages of this design is that ES modules allow renaming both imports and exports, allowing easy workarounds.

Net, the vast majority of times a developer wants to invoke a component or helper, it will be unambiguous which they are intending to invoke, and we should bias toward the best developer experience for the majority case rather than optimizing for these exceptionally rare scenarios with very easy workarounds.

My overall take is no one should ever type `<Gree` and have their editor autocomplete `<GreetingComponent` for them unless the author of the module has made a conscious and explicit choice to name the export that!